### PR TITLE
Add rollout annotation logic in application

### PR DESCRIPTION
**Pull request by ryanzhang-oss**
1.  application controller don't touch applications with this annotation
2. webhook blocks any rollout annotation to be added to an existing application

### CODE REVIEW

File: design/vela-core/rollout-design.md [hunk 1]
```diff
@@ -136,26 +136,25 @@ type RolloutPlan struct {
...
+- The applicationDeployment controller only rollout one component in the applications for now.
+- Users in general should rely on the rollout CR to do the actual rollout which means they
  shall set the `replicas` or `partition` field of the new resources to the starting value
  indicated in the [detailed rollout plan design](#rollout-plan-work-with-different-type-of-workloads).
-- 
     
  
 ## Notable implementation level details
-Here are some high level implementation decisions
+Here are some high level implementation details based on the
```
Comment (wonderflow) [inline design/vela-core/rollout-design.md:155]: on what?

File: design/vela-core/rollout-design.md [hunk 2]
```diff
@@ -165,26 +164,38 @@ upon. For example, the applicationDeployment controller will get the component f
...
-- The appDeployment controller can change the application fields. For example, it can remove all
- the conflict traits, such as HPA, during the upgrade period. 
-- Upon a successful rollout, the appDeployment controller can remove all the traits in the old
-  application and leave no pods in the workload.
-- Upon a failed rollout, the condition is not determined, it could result in an unstable state.
+#### ApplicationDeployment workflow 
+When an appDeployment is used to do application level rollout, **the target application
+is not reconciled by the application controller yet**. This is to make sure  the
+appDeployment controller has the full control of the new application from the beginning.
+We will use a pre-defined annotation "app.oam.dev/rollout" that equals to "true" to facilitate
```
Comment (wonderflow) [inline design/vela-core/rollout-design.md:171]:
```suggestion
We will use a pre-defined annotation "app.oam.dev/template" that equals to "true" to facilitate
```

I would suggest use some words like `template`/`dummy` or something that can obviously tell the CR won't really work.

File: design/vela-core/rollout-design.md [hunk 3]
```diff
@@ -165,26 +164,38 @@ upon. For example, the applicationDeployment controller will get the component f
...
+ that. We expect any system uses an appDeployment object to follow this rule.
```
Comment (wonderflow) [inline design/vela-core/rollout-design.md:172]: we can reference our apiserver design as the outside system here

File: pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go [hunk 4]
```diff
@@ -60,6 +63,15 @@ func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 		return ctrl.Result{}, err
 	}
 
+	// Check if the oam rollout annotation exists
+	if _, exist := app.GetAnnotations()[oam.AnnotationAppRollout]; exist {
```
Comment (wonderflow) [inline pkg/controller/core.oam.dev/v1alpha2/application/application_controller.go:67]: it should be behind of the deletion check in line 75~77

**Review by wonderflow — COMMENTED** [review]
The code part LGTM, please condider the annotation name

File: design/vela-core/rollout-design.md (hunk 2 above)
Comment (ryanzhang-oss) [reply design/vela-core/rollout-design.md:171]: How about "rollout-template" cause I feel like "template" alone is too general

Comment (wonderflow) [reply design/vela-core/rollout-design.md:172]: `rollout-template` is fine

**Review by wonderflow — APPROVED** [review]

**Pull request merged**